packages/contract/src/contract-new.ts

```diff
@@ -0,0 +1,273 @@
+
+    // Update transaction script with new forward params
+    this.setArguments(this.args);
+
+    return this;
+  }
+
+  private static getCallOptions(options?: CallOptions) {
+    return { fundTransaction: true, returnResult: false, ...options };
+  }
+
+  async call<T = any>(options?: CallOptionsResult): Promise<FunctionInvocationResult<T>>;
+  async call<T = any>(options?: CallOptions): Promise<T> {
+    const opts = FunctionInvocationScope.getCallOptions(options);
+    if (!this.contract.wallet) {
+      throw new Error('Wallet is required!');
+    }
+    if (opts.fundTransaction) {
+      await this.addRequiredCoins();
+    }
+    const response = await this.contract.wallet?.sendTransaction(this.transactionRequest);
+    const result = await FunctionInvocationResult.build(this, response);
+
+    return opts.returnResult ? result : result.value;
+  }
+
+  async simulate<T = any>(options?: CallOptionsResult): Promise<FunctionInvocationResult<T>>;
+  async simulate<T = any>(options?: CallOptions): Promise<T> {
+    const opts = FunctionInvocationScope.getCallOptions(options);
+    if (!this.contract.wallet) {
+      throw new Error('Wallet is required!');
+    }
+    if (opts.fundTransaction) {
+      await this.addRequiredCoins();
+    }
+    const response = await this.contract.wallet.simulateTransaction(this.transactionRequest);
+    const result = await FunctionCallResult.build(this, response);
+
+    return opts.returnResult ? result : result.value;
+  }
+
+  // simulate() {
+  //   return new FunctionInvocationResult();
+  // }
+
+  // dryRun() {
+  //   return new FunctionInvocationResult();
+  // }
+}
+
+interface Methods {
+  [key: string]: (...args: Array<unknown>) => FunctionInvocationScope;
+}
+
+export default class Contract extends AbstractContract {
+  interface!: Interface;
+  id!: string;
+  provider!: Provider | null;
+  wallet!: Wallet | null;
+  transaction?: string;
+  request?: TransactionRequest;
+  functions: Methods = {};
+
+  constructor(
+    id: string,
+    abi: JsonAbi | Interface,
+    walletOrProvider: Wallet | Provider | null = null,
+    transactionId?: string,
+    request?: TransactionRequest
+  ) {
+    super();
+    this.interface = abi instanceof Interface ? abi : new Interface(abi);
+    this.id = id;
+    this.transaction = transactionId;
+    this.request = request;
+
+    if (walletOrProvider instanceof Wallet) {
+      this.provider = walletOrProvider.provider;
+      this.wallet = walletOrProvider;
+    } else if (walletOrProvider instanceof Provider) {
+      this.provider = walletOrProvider;
+      this.wallet = null;
+    } else {
+      this.provider = null;
+      this.wallet = null;
+    }
+
+    Object.keys(this.interface.functions).forEach((name) => {
+      const fragment = this.interface.getFunction(name);
+      Object.defineProperty(this.functions, fragment.name, {
+        value: this.buildFunction(fragment),
+        writable: false,
```

Comment:
Not a blocker - Feels weird, i think we should have getter function to get `functions` and maybe make `this.functions`  private. But I'm not too opinionated on this

Reply:
Didn't find a nice way to implement the methods with types directly in the class. Mainly to support types...

Would be possible to do it by add `[key: string]: any` but will lose the power to use the contract without generated types like:
```ts
// With this type you can inform the method expects a bigint and return a bigint.
const {value} = contract.functions.method<[bigint], bigint>(1).call();
```